### DISCUSSION TITLE
Only Work days in Week Mode

### DISCUSSION THREAD

**Discussion by graldij** [kvyatkovskys/KVKCalendar#168] [Ideas]
I would need to display only workdays in the week mode (i.e. Monday-Friday). I extensively went through the documentation, but I haven't found anything yet. 
Does anyone know if this feature is possible without self-implementing it? 
Thanks, and all the best!
J

**Comment by kvyatkovskys** [comment]
Hello 👋
For what kind of view? (day/week/month/year) 🤔

**Reply by graldij** [reply]
Hi! Thanks for your answer!
In week-view!

**Reply by kvyatkovskys** [reply]
In the current version you can't display only workdays. Need to add some improvements

**Reply by graldij** [reply]
Ok don't worry, thanks anyway for all the already existing features. That would however be a good improvement in my opinion.

**Reply by nefch** [reply]
I need exactly the same. The week view should optionally only be able to show Monday to Friday, i.e. without Saturday and Sunday.

**Comment by graldij** [comment]
Hi @kvyatkovskys . I have just posted a [stackoverflow question](https://stackoverflow.com/questions/66522214/strange-behavior-of-binding-variable-in-swiftui-integrating-uikit-ios-app) for an unrelated problem. If you could take a look and (possibly) give me an advice it would be great!

**Reply by graldij** [reply]
I am developing an App for my school's courses, displaying the descriptions of the courses as well as a calendar showing the appointments. I decided to use KVKCalendar as a library for the Calendar. Developing with SwiftUI and being the KVKCalendar library a UIKit environment, I had to interface the two systems (with the struct `CalendarRepresentable`, see code). 

The app is constructed as follow: I load from the memory an `@ObservedObject` with the saved objects, pass it to some sub-views as `@Binding` in order to save/delete new courses from the course-catalogue. 
All changes made in the views are updated accordingly in all the other views, but unfortunately not in the `TimetableView`, connected with the struct `CalendarRepresentable`.

The problem: I want to update the Calendar whenever new courses are saved (e.g. added to the array `courses` of the type `[Courses]`). At the moment the struct `CalendarRepresentable` is calling the `makeUIView` as expected but the calendar is not updated anymore. In particular the saved courses in `courses` are not constantly updated but in a (apparently) inconsitent way: in some functions of the struct `CalendarRepresentable` are indeed up to date, in some other functions (and subclasses) not. See code where I show where and in what case these inconsistencies occure.

Code of the View calling the UIKit-SwiftUI interface:
import SwiftUI
import KVKCalendar

    struct TimetableView: View {
        @Binding var courses: [Courses]
        let saveAction: () -> Void
        var body: some View {
            CalendarRepresentable(courses: $courses)
                
        }
    }
    
    struct TimetableView_Previews: PreviewProvider {
        static var previews: some View {
            TimetableView(courses: .constant(Courses.data), saveAction: {})
        }
    }

Code of the interface:

    import SwiftUI
    import KVKCalendar
    import EventKit
    
    struct CalendarRepresentable: UIViewRepresentable{
        
        @Binding var courses: [Courses]
        
        var events = [Event]()
    
     
        var calendar: CalendarView = {
            print("Representable has been launched")
            var style = Style()
            //style of the calendar
            //...       
            return CalendarView(frame: UIScreen.main.bounds, style: style)
        }()
        
        
        func updateUIView(_ uiView: CalendarView, context: Context ){
            print("updateUIView has been called. Courses has \(courses.count) elements") // **Here courses has always the correct amount of elements**
            
            calendar.dataSource = context.coordinator
            calendar.delegate = context.coordinator
            calendar.reloadData()
            calendar.reloadInputViews()
            print("updateUIView is finished")
        }
        
        public func passCourses() -> [Courses]{
            print("Called passCourses with courses having \(courses.count) elements") //**Here courses has NOT the corrent amount of elements, i.e. after saving/deleting a course courses.count is different from what is printed in the previous function updateUIView**
            return courses
        }
        
        
        func makeUIView(context: Context) -> CalendarView {
            print("makeUIView has been called")
            calendar.dataSource = context.coordinator
            calendar.delegate = context.coordinator
            calendar.reloadData()
            return calendar
        }
        
        
        
        func makeCoordinator() -> Coordinator {
            print("makeCoordinator")
            return Coordinator(self)
        }
          
        
        
        
        class Coordinator: NSObject, CalendarDelegate, CalendarDataSource {
            var events = [Event]()
            var parent: CalendarRepresentable
            
            func eventsForCalendar(systemEvents: [EKEvent]) -> [Event] {
                print("eventsForCalendar called. In Coordinator \(events.count) events and \(parent.courses.count) courses") //**here courses are again not uptodate to the CalendarRepresentable' courses**
                loadEvents { (events) in
                    self.events = events
    
                }
                return self.events
            }
            
            
    
            
            init(_ parent: CalendarRepresentable){
                print("Initialize Coordinator")
                self.parent = parent
                super.init()
                
                loadEvents { (events) in
                    self.events = events
                    self.parent.calendar.reloadData()
                }
            }
            
            func eventsForCalendar() -> [Event] {
                print("eventsForCalendar without parameter")
                return events
            }
            
            
            
            func loadEvents(completion: ([Event]) -> Void) {
                
                var events = [Event]()
            
                print("loadEvents with courses having \(self.parent.courses.count) elements")//**here courses are again not uptodate to the CalendarRepresentable' courses**
    
                var i: Int=0
                for course in self.parent.passCourses() {
    
                    let isoDateStart = course.startEvent
                    let dateStartFormatter = ISO8601DateFormatter()
    
                    let isoDateEnd = course.endEvent
                    let dateEndFormatter = ISO8601DateFormatter()
                    var event = Event(ID: "\(i)")
    
                    event.start = dateStartFormatter.date(from: isoDateStart)!
                    event.end = dateEndFormatter.date(from: isoDateEnd)!
                    event.text = course.name
    
                    events.append(event)
    
                    i=i+1
                }
          
                completion(events)
            }
            
            
            
        }
        
    }

Example of console output after saving a course (done in another view) with (apparent) inconsistency:

    Representable has been launched
    updateUIView has been called. Courses has 4 elements
    eventsForCalendar called. In Coordinator 3 events and 3 courses
    loadEvents with courses having 3 elements
    Called passCourses with courses having 3 elements
    updateUIView is finished
    
Does anyone know what is happening to `courses`? Suggestions on how should I solve the issue?     
Thanks a lot for your time and help!

**Reply by kvyatkovskys** [reply]
I'll take a look later 🙂

**Reply by graldij** [reply]
Thanks! Also, if you have a better solution to add new events in the view if new courses are saved please let me know!

**Reply by graldij** [reply]
Hello @kvyatkovskys, after extensive testing I found that the problem was that somehow the coordinator had been calling with an old parent although the outerstruct had been changed. I am not sure about the internal cause of that behavior, but anyway it should now be solved: the `courses` data is now successfully passed to the coordinator. However, the view still doesn't update itself despite a change in `courses` correctly triggering `updateUIView`. Since this problem is more library-related, I will ask a question in the "issues" section hoping that someone can help me.
